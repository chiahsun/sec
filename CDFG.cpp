#include "CDFG.h"
#include <cassert>
#include <sstream>

std::string strIndent = "  ";
const std::string VarNode::VAR_BITVECTOR_DEFAULT = "0";
const std::string ConstNode::CONST_BITVECTOR_DEFAULT = "0";

std::string indent(int i = 1) {
    std::stringstream ss;
    for (int k = 0; k < i; ++k) ss << strIndent;
    return ss.str();
}

VarNode VarNode::makeInt(std::string name) {
    return VarNode(name, VAR_INT_DEFAULT);
}

VarNode VarNode::makeBitvector(std::string name) {
    return VarNode(name, VAR_BITVECTOR_DEFAULT);
}

VarNode::VarNode(std::string name, int value) 
  : _varName(name)
  , _varType(VAR_INT)
  , _varInt(value)
  , _varBitvector(VAR_BITVECTOR_DEFAULT)
  { }

VarNode::VarNode(std::string name, std::string bitvectorExpr)
  : _varName(name)
  , _varType(VAR_BITVECTOR)
  , _varInt(VAR_INT_DEFAULT)
  , _varBitvector(bitvectorExpr)
  { }
    
VarNode::VarNodeType VarNode::getType() const
  { return _varType; }

std::string VarNode::getName() const
  { return _varName; }
    
int VarNode::getInt() const
  { assert(getType() == VAR_INT); return _varInt; }
    
BitVector VarNode::getBitvector() const
  { assert(getType() == VAR_BITVECTOR); return _varBitvector; }

std::string VarNode::strPrint(int depth) const {
    std::stringstream ss;
    ss << indent(depth);
    switch (getType()) {
        case VAR_INT:
            ss << "[VAR]\n" << getName() << " int"; 
            break;
        case VAR_BITVECTOR:
            ss << "[VAR]\n" << getName() << " bitvector " << _varBitvector.size();
            break;
        default:
            break;
    }
    return ss.str();
}

std::string VarNode::strLispMode(int depth) const {
    std::stringstream ss;
    ss << indent(depth);
    switch (getType()) {
        case VAR_INT:
            ss << "(VAR " << getName() << " int)";
            break;
        case VAR_BITVECTOR:
            ss << "(VAR " << getName() << " bitvector " << _varBitvector.size() << ")";
            break;
        default:
            break;
    }    
    return ss.str();
}

ConstNode::ConstNode(int Value)
  : _constType(CONST_INT)
  , _constInt(Value)
  , _constBitvector(CONST_BITVECTOR_DEFAULT)
  {}

ConstNode::ConstNode(std::string bitvectorExpr)
  : _constType(CONST_BITVECTOR)
  , _constInt(CONST_INT_DEFAULT)
  , _constBitvector(bitvectorExpr)
  { }

ConstNode::ConstNodeType ConstNode::getType() const
  { return _constType; }

int ConstNode::getInt() const
  { return _constInt; }

BitVector ConstNode::getBitvector() const
  { return _constBitvector; }

std::string ConstNode::strPrint(int depth) const {
    std::stringstream ss;
    ss << indent(depth);
    switch (getType()) {
        case CONST_INT:
            ss << "[CONST]\n" << getInt(); 
            break;
        case CONST_BITVECTOR:
            ss << "[CONST]\n" << getBitvector();
            break;
        default:
            break;
    }
    return ss.str();
}

std::string ConstNode::strLispMode(int depth) const {
    std::stringstream ss;
    ss << indent(depth);
    switch (getType()) {
        case CONST_INT:
            ss << "(CONST " << getInt() << ")";
            break;
        case CONST_BITVECTOR:
            ss << "(CONST " << getBitvector() << ")";
            break;
        default:
            break;
    }    
    return ss.str();
}
