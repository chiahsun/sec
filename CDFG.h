#ifndef CDFG_H
#define CDFG_H

#include <iostream>
#include <string>

class Node {
public:
    virtual ~Node() {}
    void print() const {
        std::cout << strPrint(0);
    }
    virtual std::string strPrint(int depth = 0) const = 0;
    void printLispMode() const {
        std::cout << strLispMode(0);
    }
    virtual std::string strLispMode(int depth = 0) const = 0;
};


typedef std::string BitVector;

class VarNode : public Node {
public:
    enum VarNodeType { 
        VAR_INT = 0
      , VAR_BITVECTOR 
    };
private:
    const static std::string VAR_BITVECTOR_DEFAULT;
    const static int VAR_INT_DEFAULT = 0;
    const std::string _varName;
    const VarNodeType _varType;

    const int _varInt;
    const BitVector _varBitvector;
public:
    static VarNode makeInt(std::string name);
    static VarNode makeBitvector(std::string bitvector);
    VarNode(std::string name, int value);
    VarNode(std::string name, std::string bitvectorExpr);

    VarNodeType getType() const;
    std::string getName() const;
    
    int getInt() const;
    BitVector getBitvector() const;
    std::string strPrint(int depth = 0) const;
    std::string strLispMode(int depth = 0) const;

};

class ConstNode : public Node {
public:
    enum ConstNodeType {
        CONST_INT = 0
      , CONST_BITVECTOR
    };
    ConstNode(int Value);
    ConstNode(std::string bitvectorExpr);

    ConstNodeType getType() const;
    
    int getInt() const;
    BitVector getBitvector() const;

    std::string strPrint(int depth = 0) const;
    std::string strLispMode(int depth = 0) const;

private:
    const static std::string CONST_BITVECTOR_DEFAULT;
    const static int CONST_INT_DEFAULT = 0;
    const ConstNodeType _constType;
    const int _constInt;
    const BitVector _constBitvector;
};


#endif // CDFG_H
