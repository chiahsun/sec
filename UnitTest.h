#ifndef UNIT_TEST_H
#define UNIT_TEST_H

#include <iostream>

#define assertEqual(a, b, non_equal_message) do { if ((a) != (b)) { std::cerr << "TestFail : (" << __FILE__ << " : " << __LINE__ << ")\nGolden: "\
    << (a) << "\nOutput: " << (b) << "\nMessage: " << non_equal_message << "\n"; }} while(0)

#endif // UNIT_TEST_H
