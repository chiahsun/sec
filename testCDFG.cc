#include "CDFG.h"
#include "UnitTest.h"

void testVar() {
    {
        VarNode v("a", 10);
        assertEqual("a", v.getName(), "Test VarNode name");
        assertEqual(VarNode::VAR_INT, v.getType(), "Test VarNode type");
        assertEqual(10, v.getInt(), "Test VarNode Int");
        assertEqual("[VAR]\na int", v.strPrint(0), "Test VarNode print"); 
        assertEqual("(VAR a int)", v.strLispMode(0), "Test VarNode print lisp mode");
        assertEqual("(VAR aa int)", v.strLispMode(0), "Test VarNode print lisp mode");
    }
    {
        VarNode v1 = VarNode::makeInt("v1");
        VarNode v2 = VarNode::makeBitvector("v2");
        assertEqual("v1", v1.getName(), "Test make int name");
        assertEqual("v2", v2.getName(), "Test make bitvector name");
        assertEqual(VarNode::VAR_INT, v1.getType(), "Test make int type");
        assertEqual(VarNode::VAR_BITVECTOR, v2.getType(), "Test make bitvector type");
    }
    {
        VarNode v("b", "10101011");
        assertEqual("b", v.getName(), "Test VarNode name");
        assertEqual(VarNode::VAR_BITVECTOR, v.getType(), "Test VarNode type");
        assertEqual("10101011", v.getBitvector(), "Test VarNode bitvector");
        assertEqual(8, v.getBitvector().size(), "Test VarNode bitvector size");
        assertEqual("[VAR]\nb bitvector 8", v.strPrint(0), "Test Varnod print");
        assertEqual("(VAR b bitvector 8)", v.strLispMode(0), "Test VarNode lisp mode");
    }
}

void testConst() {
    {
        ConstNode v(10);
        assertEqual(ConstNode::CONST_INT, v.getType(), "Test ConstNode Type");
        assertEqual(10, v.getInt(), "Test ConstNode Int");
        assertEqual("[CONST]\n10", v.strPrint(0), "Test ConstNode print");
        assertEqual("(CONST 10)", v.strLispMode(0), "Test ConstNode lisp mode");
    }
    {
        ConstNode v("10101011");
        assertEqual(ConstNode::CONST_BITVECTOR, v.getType(), "Test ConstNode Type");
        assertEqual("10101011", v.getBitvector(), "Test ConstNode Bitvector");
        assertEqual("[CONST]\n\"10101011\"", v.strPrint(0), "Test ConstNode print");
        assertEqual("(CONST \"10101011\")", v.strLispMode(0), "Test ConstNode lisp mode");
    }
}

int main() {
    testVar();
    return 0;
}
