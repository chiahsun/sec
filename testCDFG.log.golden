TestFail : (testCDFG.cc : 12)
Golden: (VAR aa int)
Output: (VAR a int)
Message: Test VarNode print lisp mode
